[10.0 ][ADD] sale_stock_restocking_fee_invoicing: New addon to charge r…

lmignon opened the pull request:
…eturn fees in case of return of goods by a customer.

lmignon committed:
[ADD] sale_stock_picking_return_fees_invoicing: New addon to charge return fees in case of return of goods by a customer.

jbaudoux commented on sale_stock_picking_return_fees_invoicing/models/stock_move.py:
```diff
@@ -0,0 +1,38 @@
+        for move in self:
+            if not move.charge_customer_stock_return:
+                continue
+            if not move.origin_returned_move_id:
```
What happen if you make a return of a return? Shouldn't you check for is_customer_return ?

lmignon replied:
`charge_customer_stock_return` will never be True in this case. We don't need to put in place a double check

jbaudoux commented on sale_stock_picking_return_fees_invoicing/README.rst:
```diff
@@ -0,0 +1,102 @@
+is done, a new line is added into the origin sale order to charge fees. The
+quantity charged is always 1 whatever the number of items of the same product
+returned.
+
```
Worth adding that in your implementation 1 fee line is added per returned SO line. And not 1 fee line per return.

lmignon replied:
No, we add a line by return not by so line. If you purchase 2 product and return the product 1 item at a time (1 pickings) you'll be charged twice

jbaudoux commented on sale_stock_picking_return_fees_invoicing/models/sale_order.py:
```diff
@@ -0,0 +1,57 @@
+                    "contact you administrator."
+                )
+            )
+        description = _("Return fees for %s %s %s") % (
```
This should be in the language of the customer

lmignon replied:
indeed

jbaudoux reviewed (commented):
Some question/remarks

lmignon committed:
[ADD] sale_stock_picking_return_fees_invoicing: The name of the line created for return fees into the SO must be into the customer lang

lmignon committed:
[FIX] sale_stock_picking_return_fees_invoicing: avoid integrity error

pedrobaeza commented:
Please avoid using plural in module name as stated in the contribution guidelines: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#modules

lmignon commented:
@pedrobaeza I know but in this case AFAIK the meaning of the plural form is not the same as the singular form.
fee =~ cost
fees =~ charge 
IMO the plural form is the one appropriate to describe the functionality. Perhaps other people with a better level of english can validate/invalidate the correct use of the term fees.

pedrobaeza commented:
OK, I have no other better suggestion for now.

lmignon commented:
@pedrobaeza I agree with the remark and I am ready to change if my understanding is wrong or if another term is more appropriate. Finding a good module name is something I always find difficult. For a while I had also thought about `customer_return_fees_invoicing`.

pedrobaeza commented:
![imagen](https://user-images.githubusercontent.com/7165771/93199625-e0d84100-f74e-11ea-99c2-d87e2087b3bd.png)

:rofl:

lmignon committed:
[FIX] sale_stock_picking_return_fees_invoicing: Wrong domain

gdgellatly commented:
Maybe use the word restocking somwhere. In the English speaking world that is what we call them.

A Restocking Fee by the way, not plural. https://dictionary.cambridge.org/dictionary/english/restocking-fee

gdgellatly commented:
Hi I did a quick review. For the most part the structure is as expected, but in the countries and companies I have worked (so all high value B2B - not retail, UK/AU/NZ), restocking fees behave pretty much the same as order deposits. A percentage of the value of goods returned. 

Usually how we handle them in Odoo, we just put a discount on the credit note line for each product line.

Not blocking, just feedback, maybe for a todo or something to basically emulate the sale order deposit interface because fixed charge just won't work for a very large number of companies/countries.

lmignon commented:
Hi @gdgellatly 
Thank you for your review and feedback.
@pedrobaeza @gdgellatly What do you thing about 'restocking_fee_invoicing'?

I will also add a comment in the roadmap describing a potential evolution of the module so that it can support different return policy. Actually we support a fixed price per returned products line, but it will be nice to support a percentage of the value of the returned products.

pedrobaeza commented:
I would still keep the `sale_stock` part (so `sale_stock_restocking_fee_invoicing`) for having the technical prefix for "grouping" the module conveniently. OK for the roadmap. Thanks for this.

lmignon commented:
> I would still keep the `sale_stock` part (so `sale_stock_restocking_fee_invoicing`) for having the technical prefix for "grouping" the module conveniently. OK for the roadmap. Thanks for this.

OK I'll do the changes

lmignon committed:
[FIX] sale_stock_picking_return_fees_invoicing: Rename to sale_stock_restocking_fee_invoicing and use better naming into the code

lmignon commented:
@pedrobaeza @gdgellatly The module has been renamed and the naming into the source files adapted. In the same time I renamed the name of the method called to charge restocking fee into the sale_order to be agnostic of the invoicing policy applied for restocking fees github.com/OCA/account-invoicing/pull/784/files#diff-0b351e68bac2442912f5344d57a56478R66

pedrobaeza commented on sale_stock_restocking_fee_invoicing/data/product_product.xml:
```diff
@@ -0,0 +1,12 @@
+        <field name="list_price">100</field>
+        <field name="type">service</field>
+        <field name="sale_ok" eval="False"/>
+        <field name="invoice_policy">order</field>
```
Add `<field name="company_id" eval="False"/>` for making multi-company compliant by default.

pedrobaeza commented on sale_stock_restocking_fee_invoicing/models/res_company.py:
```diff
@@ -0,0 +1,33 @@
+        "product.product",
+        string="Restocking Fee Product",
+        ondelete="restrict",
+        default=lambda a: a._default_restocking_fee_product_id(),
```
This default will be assigned before data is loaded, so it will False. You can use a post-init hook for filling it.

pedrobaeza commented on sale_stock_restocking_fee_invoicing/models/stock_move.py:
```diff
@@ -0,0 +1,38 @@
+    def action_done(self):
+        result = super(StockMove, self).action_done()
+        move_ids_to_charge_by_so = defaultdict(list)
+        for move in self:
+            if not move.charge_restocking_fee:
+                continue
+            if not move.origin_returned_move_id:
+                continue
```
Isn't shorter this?:
```suggestion
        for move in self.filtered(lambda x: x.charge_restocking_fee and x.origin_returned_move_id):
```

lmignon replied:
IMO, the readability is better without filtered. In this case, Moreover in this case, We don't need to reduce the recordset on which the process is done.

pedrobaeza replied:
Well, at least join both conditions:
```suggestion
        for move in self:
            if not move.charge_restocking_fee or not move.origin_returned_move_id:
                continue
```

gdgellatly replied:
Matter of style I guess. Personally I'd just go with

```python
if move.charge.. and move.origin..:
 do_stuff()
```
I get sometimes it works doing a continue if there is a massive block of code underneath with more branches, but it is 2 lines.

pedrobaeza replied:
Yeah, indeed this is not blocking. I don't like that style though, as it implies an extra level of indentation for your main code, although it avoids uncovered lines in tests. We never have the ideal one!

houssine78 commented:
> 
> 
> ![imagen](https://user-images.githubusercontent.com/7165771/93199625-e0d84100-f74e-11ea-99c2-d87e2087b3bd.png)
> 
> 🤣

https://thecodinglove.com/when-im-searching-for-a-meaningful-variable-name

houssine78 commented:
> 
> 
> ![imagen](https://user-images.githubusercontent.com/7165771/93199625-e0d84100-f74e-11ea-99c2-d87e2087b3bd.png)
> 
> 🤣

![imagen](https://thecodinglove.com/content/028/searching_meaningful_variable_name2.png)

lmignon committed:
[IMP] sale_stock_restocking_fee_invoicing: Initialize restocking feee product on companies at install
Use a post_init_hook to initialize this value since the default value provided by xml data is not yet loaded when the default method is called on res.company

lmignon committed:
[IMP] sale_stock_restocking_fee_invoicing: Makes restocking product saleable and available to all companies

lmarion-source committed:
[IMP] sale_restocking_fee_invoicinf: Add hook to filter out some product where creating restocking fee

lmignon commented:
Closing since it will never be merged....
